Refresh.GameServer/Services/OAuth/OAuthClient.cs

```diff
@@ -0,0 +1,232 @@
+
+            throw new Exception($"Unexpected error {errorResponse.Error} when acquiring token! Description: {errorResponse.ErrorDescription}, URI: {errorResponse.ErrorUri}");
+        }
+
+        if (!result.IsSuccessStatusCode)
+            throw new Exception($"Acquiring token failed, server returned status code {result.StatusCode}");
+
+        if ((result.Content.Headers.ContentLength ?? 0) == 0)
+            throw new Exception("Acquiring token failed, request returned no response");
+        
+        OAuth2AccessTokenResponse response = result.Content.ReadAsJson<OAuth2AccessTokenResponse>()!;
+
+        // Case insensitive according to spec
+        if (!response.TokenType.Equals("bearer", StringComparison.InvariantCultureIgnoreCase))
+            throw new Exception("Non-bearer tokens are currently unsupported.");
+        
+        return response;
+    }
+
+    /// <summary>
+    /// Refreshes the passed OAuthTokenRelation, acquiring a new access token
+    /// </summary>
+    /// <param name="database">The database context associated with the token</param>
+    /// <param name="token">The token to refresh</param>
+    /// <param name="timeProvider">The time provider associated with the request</param>
+    /// <returns>Whether the refresh succeeded, if failed, assume the token is invalid and authorization has been revoked.</returns>
+    public bool RefreshToken(GameDatabaseContext database, OAuthTokenRelation token, IDateTimeProvider timeProvider)
+    {
+        // If we have no refresh token, then always fail the token refresh
+        if (token.RefreshToken == null)
+            return false;
+        
+        HttpResponseMessage result = this.Client.PostAsync(this.TokenEndpoint, new FormUrlEncodedContent([
+            new KeyValuePair<string, string>("grant_type", "refresh_token"),
+            new KeyValuePair<string, string>("refresh_token", token.RefreshToken),
+            new KeyValuePair<string, string>("client_id", this.ClientId),
+            new KeyValuePair<string, string>("client_secret", this.ClientSecret),
+        ])).Result;
+
+        // https://datatracker.ietf.org/doc/html/rfc6749#section-5.2
+        if (result.StatusCode == BadRequest)
+        {
+            OAuth2ErrorResponse errorResponse = result.Content.ReadAsJson<OAuth2ErrorResponse>()!;
+
+            //Special cased error for when the refresh token is invalid
+            if (errorResponse.Error == "invalid_grant")
+                return false;
+
+            throw new Exception($"Unexpected error {errorResponse.Error} when refreshing token! Description: {errorResponse.ErrorDescription}, URI: {errorResponse.ErrorUri}");
+        }
+
+        if (!result.IsSuccessStatusCode)
+            throw new Exception($"Refreshing token failed, server returned status code {result.StatusCode}");
+
+        if ((result.Content.Headers.ContentLength ?? 0) == 0)
+            throw new Exception("Refreshing token failed, request returned no response");
```

Comment:
This error checking block could be extracted since it entirely duplicates the one in `AcquireTokenFromAuthorizationCode` and `RevokeToken`.